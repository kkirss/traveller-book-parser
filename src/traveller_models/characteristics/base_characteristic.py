from enum import Enum
from typing import ClassVar, Literal

from pydantic import BaseModel, Field, computed_field

from traveller_models.entity_types import EntityType


class CharacteristicType(str, Enum):
    STR = "STR"
    DEX = "DEX"
    END = "END"

    INT = "INT"
    EDU = "EDU"
    SOC = "SOC"

    LCK = "LCK"
    INF = "INF"


class BaseCharacteristic(BaseModel):
    entity_type: Literal[EntityType.CHARACTERISTIC] = EntityType.CHARACTERISTIC
    characteristic_type: CharacteristicType = Field(repr=True)

    level: int | None = Field(default=None, repr=True)

    NAME: ClassVar[str] = NotImplemented

    @computed_field
    @property
    def name(self) -> str:
        return self.NAME
