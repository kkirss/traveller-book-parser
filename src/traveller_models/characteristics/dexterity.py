from typing import ClassVar, Literal

from .base_characteristic import BaseCharacteristic, CharacteristicType


class Dexterity(BaseCharacteristic):
    characteristic_type: Literal[CharacteristicType.DEX
                                 ] = CharacteristicType.DEX

    NAME: ClassVar[str] = "Dexterity"
