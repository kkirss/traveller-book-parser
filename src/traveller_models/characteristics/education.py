from typing import ClassVar, Literal

from .base_characteristic import BaseCharacteristic, CharacteristicType


class Education(BaseCharacteristic):
    characteristic_type: Literal[CharacteristicType.EDU]

    NAME: ClassVar[str] = "Education"
