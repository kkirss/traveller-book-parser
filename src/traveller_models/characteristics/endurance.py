from typing import ClassVar, Literal

from .base_characteristic import BaseCharacteristic, CharacteristicType


class Endurance(BaseCharacteristic):
    characteristic_type: Literal[CharacteristicType.END] = CharacteristicType.END

    NAME: ClassVar[str] = "Endurance"
