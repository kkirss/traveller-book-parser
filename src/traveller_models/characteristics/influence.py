from typing import ClassVar, Literal

from .base_characteristic import BaseCharacteristic, CharacteristicType


class Influence(BaseCharacteristic):
    characteristic_type: Literal[CharacteristicType.INF]

    NAME: ClassVar[str] = "Influence"
