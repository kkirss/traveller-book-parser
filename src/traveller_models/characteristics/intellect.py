from typing import ClassVar, Literal

from .base_characteristic import BaseCharacteristic, CharacteristicType


class Intellect(BaseCharacteristic):
    characteristic_type: Literal[CharacteristicType.INT
                                 ] = CharacteristicType.INT

    NAME: ClassVar[str] = "Intellect"
