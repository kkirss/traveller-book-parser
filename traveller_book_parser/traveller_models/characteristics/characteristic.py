from typing import Annotated, Any

from pydantic import Field, RootModel

from .dexterity import Dexterity
from .education import Education
from .endurance import Endurance
from .influence import Influence
from .intellect import Intellect
from .luck import Luck
from .social_standing import SocialStanding
from .strength import Strength

Characteristic = Annotated[
    Strength
    | Dexterity
    | Endurance
    | Intellect
    | Education
    | SocialStanding
    | Influence
    | Luck,
    Field(discriminator="characteristic_type"),
]


def get_characteristic_model(**kwargs: dict[str, Any]) -> Characteristic:
    return RootModel[Characteristic](**kwargs).root
