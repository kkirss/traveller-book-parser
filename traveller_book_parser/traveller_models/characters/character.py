from typing import Any, Literal

from pydantic.v1 import Field

from traveller_book_parser.traveller_models.characteristics.characteristic import (
    Characteristic,
)
from traveller_book_parser.traveller_models.skills.skill import Skill
from traveller_book_parser.traveller_models.trav_object_base import TravObjectBase
from traveller_book_parser.traveller_models.trav_object_types import TravObjectType


class Character(TravObjectBase):
    """Character."""

    type: Literal[TravObjectType.CHARACTER] = Field(
        repr=False, default=TravObjectType.CHARACTER
    )

    characteristics: list[Characteristic] = Field(default_factory=list)
    skills: list[Skill] = Field(default_factory=list)


def create_character(**kwargs: Any) -> Character:  # noqa: ANN401
    """Create Character model instance."""
    return Character(**kwargs)
