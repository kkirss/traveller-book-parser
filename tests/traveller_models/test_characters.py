from syrupy.assertion import SnapshotAssertion

from traveller_book_parser.traveller_models.characteristics.characteristic import (
    CharacteristicType,
)
from traveller_book_parser.traveller_models.characters.character import create_character


def test_character(snapshot: SnapshotAssertion):
    input_data = {
        "type": "character",
        "name": "John Doe",
        "characteristics": [
            {
                "characteristic_type": CharacteristicType.STR.value,
                "name": "Strength",
                "level": 7,
            },
            {
                "characteristic_type": CharacteristicType.DEX.value,
                "name": "Dexterity",
                "level": 5,
            },
        ],
        "skills": [{"name": "Admin", "level": 1}],
    }
    model = create_character(**input_data)

    assert model.model_dump(exclude_unset=True) == snapshot
    assert model.trav_id == "character:John Doe"


def test_character_with_colon_in_name(snapshot: SnapshotAssertion):
    input_data = {
        "type": "character",
        "name": "hello:world",
        "characteristics": [],
        "skills": [],
    }
    model = create_character(**input_data)

    assert model.model_dump(exclude_unset=True) == snapshot
    assert model.trav_id == "character:hello-world"
