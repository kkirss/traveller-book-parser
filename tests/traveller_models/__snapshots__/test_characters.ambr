# serializer version: 1
# name: test_character
  dict({
    'characteristics': list([
      dict({
        'characteristic_type': <CharacteristicType.STR: 'STR'>,
        'level': 7,
        'name': 'Strength',
        'trav_id': 'characteristic:Strength',
      }),
      dict({
        'characteristic_type': <CharacteristicType.DEX: 'DEX'>,
        'level': 5,
        'name': 'Dexterity',
        'trav_id': 'characteristic:Dexterity',
      }),
    ]),
    'name': 'John Doe',
    'skills': list([
      dict({
        'level': 1,
        'name': 'Admin',
        'trav_id': 'skill:Admin:1',
      }),
    ]),
    'trav_id': 'character:John Doe',
    'type': <TravObjectType.CHARACTER: 'character'>,
  })
# ---
# name: test_character_with_colon_in_name
  dict({
    'characteristics': list([
    ]),
    'name': 'hello:world',
    'skills': list([
    ]),
    'trav_id': 'character:hello-world',
    'type': <TravObjectType.CHARACTER: 'character'>,
  })
# ---
